### PR TITLE
Encode the rhythm tapped by the user as a key

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: -
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository an encoding of the rhythm tapped by the user as a unique key by using K-means clustering

## Why
This Pull Request is needed because we need a key, derived by the rhythm, to encode / decode the Bluetooth information

## How
This feature can be viewed/tested within the project by running the tests

## Alternative implementation
Other implementations that I've have considered are 

## Notes

3 Cucumber tests are failing, but I have no idea why. I would greatly appreciate it if you can take a look!

### CODE REVIEW

File: app/src/androidTest/assets/features/create_rhythm.feature [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Feature: Create a rhythm
+When you create a rhythm
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/create_rhythm.feature]: You're missing indentation here

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/first_time_use.feature [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Feature: First time use
```
Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/first_time_use.feature]: Why did you move the feature files from ~/assets/features to ~/test

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 3]
```diff
@@ -167,13 +193,165 @@ public void onNewDataReceivedEvent(final NewDataReceivedEvent event) {
...
+                clusterCenterMean1 = newClusterCenters.getLeft();
+                clusterCenterMean2 = newClusterCenters.getRight();
+            }
+
+            final long breakpoint = this.getBreakpoint();
+            this.intervals = getIntervals(taps);
+            return this.generateKey(breakpoint);
+        }
+
+        private long makeAndReturnFirstClusterMean() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Missing JavaDoc, also applies to other methods in this class

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 4]
```diff
@@ -117,9 +127,25 @@ public void onStartRecordingClick(final View v) {
...
+        this.doneButton.setVisibility(View.VISIBLE);
+        if (this.taps.size() < MINIMUM_TAPS) {
+            this.taps.clear();
+            this.doneButton.setVisibility(View.GONE);
+            new SweetAlertDialog(this, SweetAlertDialog.ERROR_TYPE)
+                    .setTitleText(this.getString(R.string.too_few_taps_title))
+                    .setContentText(this.getString(R.string.too_few_taps_description))
+                    .setConfirmText(this.getString(R.string.dialog_ok))
+                    .show();
+            /*new SweetAlertDialog(this, SweetAlertDialog.ERROR_TYPE)
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Can you remove commented code

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 5]
```diff
@@ -80,20 +89,20 @@ public void onTapClick(final View v) {
...
 
-            LOGGER.info("Sending my profile with name: " + myProfile.getName() + ", public key: "
-                    + myProfile.getPublicKey().toString());
+            LOGGER.info("Sending my profile with name: {}, public key: {}", myProfile.getName(), myProfile.getPublicKey());
             final Contact myProfileAsContact = new Contact(myProfile.getName(), new SimpleKey("simplekey", "73890ien"));
-            this.serviceLocator.getBluetoothHandler().send(myProfileAsContact);
-        } catch (IOException e) {
-            LOGGER.error("Could not send my contact to other device " + e.getMessage());
+            final int encryptionKey = new RhythmCreateActivity.KMeansClusterHelper().getEncryptionKey(this.taps);
+            this.bluetoothHandler.send(myProfileAsContact, encryptionKey);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Can you add this to the visual verification as well?

File: app/src/androidTest/assets/features/create_rhythm.feature [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    And I press the stop button
+    Then I get an error that I tapped not enough times
+
+    Examples:
+      | tap_count |
+      | 0         |
+      | 1         |
+      | 2         |
+
+  Scenario: Tapping enough times
```
Comment (TheBonheurs) [inline app/src/androidTest/assets/features/create_rhythm.feature]: enough is not very well defined i think, you now use 10 but in the strings.xml you say at least 3 times

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java [hunk 7]
```diff
@@ -57,19 +56,62 @@ public void dataReceived(final byte[] bytes) {
...
              ObjectOutputStream oos = new ObjectOutputStream(bos)) {
             oos.writeObject(new DataWrapper(object));
             oos.flush();
-            bytes = bos.toByteArray();
+            return bos.toByteArray();
+        } catch (final IOException e) {
+            LOGGER.error("Couldn't serialize object", e);
+            throw new SerializationException(e);
         }
-        send(bytes);
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java]: Why are you making these changes in this pr, this will lead to unnecessary conflicts with nfc pr

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 3 above)
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: I get an error usually from checkstyle or pmd if i miss javadoc

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 8]
```diff
@@ -24,38 +28,43 @@
...
-import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import cn.pedant.SweetAlert.SweetAlertDialog;
 
 /**
  * The RhythmCreateActivity is an Activity where you can tap a sequence.
  */
-@SuppressFBWarnings(value = "BC_BAD_CAST_TO_CONCRETE_COLLECTION")
 //List is cast to an ArrayList, but that is needed to put in an intent.
-@SuppressWarnings("PMD.LooseCoupling")
+@SuppressWarnings({"PMD.LooseCoupling", "InstanceVariableMayNotBeInitialized"})
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: add a comment fopr the second suppression

File: app/src/androidTest/assets/features/create_rhythm.feature (hunk 6 above)
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/create_rhythm.feature]: Actually, this is a perfect opportunity to use a Scenario outline, testing multiple different amounts of taps bigger then 3

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/first_time_use.feature (hunk 2 above)
Comment (jverbraeken) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/first_time_use.feature]: Oops, that was because I had to rerun that cucumber test quite a few times and I didn't want to wait for those other tests

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 5 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Yes, in another PR

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 3 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: @TheBonheurs That's right. Officially you don't need to document private stuff with javadoc because private methods and fields are not visible in the documentation that the javadoc generates. In short, you're writing documentation that won't be visible in the documentation.

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java (hunk 7 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java]: Yeah, maybe I shouldn't have done that. I did it because Intellij complained in another file that I didn't use lambda's, so I added retrolambda to solve that warning, but then I got 2 new warnings in 2 other files that also should use retrolambda, so I fixed them. This was one of them. You can safely delete this code in the nfc pr

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 5 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: 👍

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 9]
```diff
@@ -170,39 +184,6 @@ public void onStopRecordingClick(final View v) {
...
-        super.onStart();
-        this.serviceLocator.registerToEventBus(this);
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    protected void onStop() {
-        this.serviceLocator.unregisterFromEventBus(this);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Don't you want to `unregisterFromEventBus`?

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: onStop was duplicate

File: app/src/androidTest/assets/features/first_time_use.feature [hunk 10]
```diff
@@ -1,7 +0,0 @@
-Feature: First time use
```
Comment (ericcornelissen) [inline app/src/androidTest/assets/features/first_time_use.feature]: Wait 😕 when and why has this been removed?

Comment (jverbraeken) [inline app/src/androidTest/assets/features/first_time_use.feature]: It's now called welcome.feature and WelcomeSteps

Comment (ericcornelissen) [inline app/src/androidTest/assets/features/first_time_use.feature]: Ow, my bad 😅
